### PR TITLE
compiles all except some of the tests

### PR DESCRIPTION
**Pull request by jtrmal**
The tests algo_test_log and algo_test_power does not compile because:
```
7>C:\Users\jtrmal\Documents\openfst\src\include\fst/shortest-path.h(151): error C2338: Weight must have path property.
7>C:\Users\jtrmal\Documents\openfst\src\include\fst/shortest-path.h(429): note: see reference to function template instantiation 'bool fst::internal::SingleShortestPath<fst::LogArc,fst::AutoQueue<StateId>,fst::AnyArcFilter<Arc>>(const fst::Fst<fst::LogArc> &,std::vector<fst::LogWeightTpl<float>,std::allocator<_Ty>> *,const fst::ShortestPathOptions<Arc,fst::AutoQueue<StateId>,fst::AnyArcFilter<Arc>> &,fst::ArcTpl<fst::LogWeight>::StateId *,std::vector<std::pair<int,::size_t>,std::allocator<std::pair<int,::size_t>>> *)' being compiled
```
that also relates my changes src/script/shortest-path.cc in this PR (had to comment it out, as it was causing the same static_assert failure). I'm thinking if this is related to our changes .w.r.t. enable_if.

Another error is when compiling the test weight_test (weights_test.cc):
```
1>C:\Users\jtrmal\Documents\openfst\src\include\fst/set-weight.h(197): error C3646: 'iter_': unknown override specifier
1>C:\Users\jtrmal\Documents\openfst\src\include\fst/set-weight.h(198): note: see reference to class template instantiation 'fst::SetWeightIterator<SetWeight_>' being compiled
```

### CODE REVIEW

**Comment by jtrmal** [comment]
I have changed the static_assert into run-time errors, as it was in the previous versions
hopefully that won't break anything.

**Comment by jtrmal** [comment]
so far, I haven't been able to resolve the other issue (iter_)
relates to
```
  typename decltype(Weight::rest_)::const_iterator iter_;
```
in set-weight.h

File: src/script/shortest-path.cc [hunk 1]
```diff
@@ -19,8 +19,8 @@ void ShortestPath(const FstClass &ifst, MutableFstClass *ofst,
 }
 
 REGISTER_FST_OPERATION(ShortestPath, StdArc, ShortestPathArgs);
-//REGISTER_FST_OPERATION(ShortestPath, LogArc, ShortestPathArgs);
-//REGISTER_FST_OPERATION(ShortestPath, Log64Arc, ShortestPathArgs);
+REGISTER_FST_OPERATION(ShortestPath, LogArc, ShortestPathArgs);
+REGISTER_FST_OPERATION(ShortestPath, Log64Arc, ShortestPathArgs);
```
Comment (kkm000) [inline src/script/shortest-path.cc]: LGTM, but I am wondering why did you have to comment these lines back in the next commit in this MR. Is this intentional?

The log weight naturally does not have the path property... Seems like MSVC is too eager to evaluate the condition in the static_assert again, like it did with the `enable_if`!

**Comment by kkm000** [comment]
> `typename decltype(Weight::rest_)::const_iterator iter_;`

This is a problem with the syntactic MSVC parser. What I did was wrap declspec into *just anything* that would return it as a type. I could have written my own little class
```c++
template<class T>
struct Wrap {
  typedef T type;
}
```
and then use it like
```c++
typename Wrap<decltype(Weight::rest_)>::type::const_iterator iter_;
```
But what I did instead just borrowed a library template that already does that in our case:
```c++
typename std::remove_reference<decltype(Weight::rest_)>::type::const_iterator iter_;
```
should compile and be identical to the original semantics, as `decltype()` cannot be a reference. This is a hack, naturally.

File: src/script/shortest-path.cc (hunk 1 above)
Comment (kkm000) [reply src/script/shortest-path.cc]: Feel free to merge if that's intended. Are you able to merge in this repo?

**Comment by jtrmal** [comment]
yes, that remove_reference trick has helped.  I won't claim I knew what I'm
doing, though :)
y.

**Comment by kkm000** [comment]
`std::remove_reference<decltype(x)>::type` is just a less than pretty way to write `decltype(x)` :)

**Comment by jtrmal** [comment]
I think the idea originally was to return an error in case the semiring
does not have the right properties.
Using the enable_if construct, they would select of a different
implementation of the ShortestPath functions(which returns error directly),
so the static_assert would never be triggered for the usual semirings.
But when we replaced the enable_if constructs by the if() statements, these
asserts got into the compilation path again.
I'm not sure what they plan to do, anyway, so I just returned the
REGISTER_FST_OPERATION statements back to keep changes minimal
y.

**Comment by kkm000** [comment]
> I'm not sure what they plan to do, anyway, so I just returned the REGISTER_FST_OPERATION statements back to keep changes minimal

Ah, so *they* had them commented out, correct?

**Comment by jtrmal** [comment]
No, they kept them enabled. I am really confused what was their intention,
as those semirings do not have path property.
y.

**Comment by kkm000** [comment]
I do not really understand whither they are going. One idea just to have no member function a class if a class template is specialized with a class (think Weight) that does not have a certain property known at compile time. That makes sense for C++ (to give a compile error instead of a run-time error), but that does not work with the scripting code (a binary should still compile somehow, but produce a *runtime* error when given an FST file with the weight type lacking that property).

C++ templating is a very powerful thing, and this problem is perhaps solvable. But I am afraid of the moment they solve it though, and we'll have to compile the port...

So is this good to go, merge? Does not it compile with your changes from static assert  to runtime error if you uncomment these lines?

**Comment by jtrmal** [comment]
yep, will merge

**Pull request merged**